File: discord-jockey/src/util/logger.rb
# typed: true

require "logger"

class Logger
  def self.init
    @logger = ::Logger.new($stdout)
    @logger.progname = "discord-client"
    @logger.formatter = proc do |severity, datetime, progname, msg|
      "[#{progname}] #{severity}: #{msg}\n"
    end
    $stdout.sync = true
    $stderr.sync = true
  end

  def self.info(msg, **attrs)
    @logger.info(format(msg, attrs))
  end

  def self.warn(msg, **attrs)
    @logger.warn(format(msg, attrs))
  end

  def self.error(msg, **attrs)
    @logger.error(format(msg, attrs))
  end

  def self.format(msg, attrs)
    return msg if attrs.empty?

    formatted = attrs.each_with_index.map do |(k, v), i|
      i < attrs.size - 1 ? "#{k}: #{v}," : "#{k}: #{v}"
    end.join(" ")

    "#{msg} { #{formatted} }"
  end
  private_class_method :format
end

File: discord-jockey/src/util/logger.ts
type Metadata = Record<string, unknown>;

function format(level: string, message: string, metadata?: Metadata): string {
  const entries = Object.entries(metadata ?? {});
  const meta =
    entries.length > 0
      ? " {" + entries.map(([k, v]) => ` ${k}: ${v}`).join(",") + " }"
      : "";
  return `[discord-jockey] ${level}: ${message}${meta}`;
}

export const logger = {
  info(message: string, metadata?: Metadata): void {
    console.log(format("INFO", message, metadata));
  },
  error(message: string, metadata?: Metadata): void {
    console.error(format("ERROR", message, metadata));
  },
  warn(message: string, metadata?: Metadata): void {
    console.warn(format("WARN", message, metadata));
  },
};
